File: plugin/src/main/java/org/godot/admob/lite/AdLogic.java
package org.godot.admob.lite;

import android.app.Activity;
import android.view.Gravity;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;

import com.google.android.gms.ads.AdError;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdSize;
import com.google.android.gms.ads.AdView;
import com.google.android.gms.ads.FullScreenContentCallback;
import com.google.android.gms.ads.LoadAdError;
import com.google.android.gms.ads.MobileAds;
import com.google.android.gms.ads.interstitial.InterstitialAd;
import com.google.android.gms.ads.interstitial.InterstitialAdLoadCallback;
import com.google.android.gms.ads.rewarded.RewardedAd;
import com.google.android.gms.ads.rewarded.RewardedAdLoadCallback;

import java.util.HashMap;
import java.util.Map;

public class AdLogic {
    private final Activity activity;
    private final FrameLayout container;

    // Maps to store loaded ads by their IDs
    private final Map<String, AdView> loadedBanners = new HashMap<>();
    private final Map<String, InterstitialAd> loadedInterstitials = new HashMap<>();
    private final Map<String, RewardedAd> loadedRewarded = new HashMap<>();

    // Interface for callbacks
    public interface AdLoadCallback {
        void onResult(boolean success);
    }

    public interface InterstitialShowCallback {
        void onResult(int event);
    }

    public interface RewardedShowCallback {
        void onResult(int event);
    }

    public interface RewardedRewardCallback {
        void onReward(int amount, String type);
    }

    public AdLogic(Activity act, FrameLayout container) {
        this.activity = act;
        this.container = container;
    }

    public void initialize() {
        MobileAds.initialize(activity);
    }

    // Banner Logic

    public void loadBanner(String id, int size, int position) {
        destroyBanner(id); // Destroy any existing banner with the same ID

        AdView newBanner = new AdView(activity);
        if (BuildConfig.DEBUG) { 
            newBanner.setAdUnitId("ca-app-pub-3940256099942544/6300978111"); // Test Ad Unit ID
        } else {
            newBanner.setAdUnitId(id);
        }

        // Set the AdSize based on the 'size' parameter
        if (size > 40) {
            newBanner.setAdSize(AdSize.getCurrentOrientationAnchoredAdaptiveBannerAdSize(activity, size));
        } else {
            switch (size) {
                case 0:
                    newBanner.setAdSize(AdSize.BANNER);
                    break;
                case 1:
                    newBanner.setAdSize(AdSize.LARGE_BANNER);
                    break;
                case 2:
                    newBanner.setAdSize(AdSize.MEDIUM_RECTANGLE);
                    break;
                case 3:
                    newBanner.setAdSize(AdSize.FULL_BANNER);
                    break;
                case 4:
                    newBanner.setAdSize(AdSize.LEADERBOARD);
                    break;
            }
        }

        // Determine the gravity based on the 'position' parameter
        int gravity = switch (position) {
            case 0 -> Gravity.TOP | Gravity.CENTER_HORIZONTAL;
            case 1 -> Gravity.CENTER;
            case 2 -> Gravity.BOTTOM | Gravity.CENTER_HORIZONTAL;
            default ->
                    Gravity.TOP | Gravity.CENTER_HORIZONTAL; // Default to top if an invalid position is provided
        };

        // Create a new FrameLayout.LayoutParams with the determined gravity
        FrameLayout.LayoutParams layoutParams = new FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.MATCH_PARENT, // Width
                FrameLayout.LayoutParams.WRAP_CONTENT, // Height
                gravity
        );

        // Add the new banner to the container with the specified layout parameters
        container.addView(newBanner, layoutParams);

        AdRequest adRequest = new AdRequest.Builder().build();
        newBanner.loadAd(adRequest);
        loadedBanners.put(id, newBanner);
    }

    public void destroyBanner(String id) {
        AdView banner = loadedBanners.get(id);
        if (banner != null) {
            // Remove banner from view hierarchy.
            ViewGroup parentView = (ViewGroup) banner.getParent();
            if (parentView != null) {
                parentView.removeView(banner);
            }
            banner.destroy();
            loadedBanners.remove(id);
        }
    }

    // Interstitial Logic

    public void loadInterstitial(String id, final AdLoadCallback callback) {
        String adId = "ca-app-pub-3940256099942544/1033173712"; // Test Ad Unit ID
        if (!BuildConfig.DEBUG) { // Assuming BuildConfig.DEBUG is false for release
            adId = id;
        }

        InterstitialAd.load(
                activity,
                adId,
                new AdRequest.Builder().build(),
                new InterstitialAdLoadCallback() {
                    @Override
                    public void onAdLoaded(@NonNull InterstitialAd ad) {
                        callback.onResult(true);
                        loadedInterstitials.put(id, ad);
                    }

                    @Override
                    public void onAdFailedToLoad(@NonNull LoadAdError adError) {
                        callback.onResult(false);
                        loadedInterstitials.remove(id);
                    }
                }
        );
    }

    public void releaseInterstitial(String id) {
        loadedInterstitials.remove(id);
    }

    public void showInterstitial(String id, final InterstitialShowCallback callback) {
        InterstitialAd interstitialAd = loadedInterstitials.get(id);
        if (interstitialAd == null) {
            callback.onResult(0); // Ad not loaded
            return;
        }

        interstitialAd.setFullScreenContentCallback(new FullScreenContentCallback() {
            @Override
            public void onAdFailedToShowFullScreenContent(@NonNull AdError adError) {
                callback.onResult(0); // Failed to show
                loadedInterstitials.remove(id);
            }

            @Override
            public void onAdDismissedFullScreenContent() {
                callback.onResult(1); // Dismissed
                loadedInterstitials.remove(id);
            }

            @Override
            public void onAdShowedFullScreenContent() {
                callback.onResult(2); // Showed
            }

            @Override
            public void onAdImpression() {
                callback.onResult(3); // Impression
            }

            @Override
            public void onAdClicked() {
                callback.onResult(4); // Clicked
            }
        });
        interstitialAd.show(activity);
    }

    // Rewarded Logic

    public void loadRewarded(String id, final AdLoadCallback callback) {
        String adId = "ca-app-pub-3940256099942544/5224354917"; // Test Ad Unit ID
        if (!BuildConfig.DEBUG) { // Assuming BuildConfig.DEBUG is false for release
            adId = id;
        }

        RewardedAd.load(
                activity,
                adId,
                new AdRequest.Builder().build(),
                new RewardedAdLoadCallback() {
                    @Override
                    public void onAdLoaded(@NonNull RewardedAd ad) {
                        callback.onResult(true);
                        loadedRewarded.put(id, ad);
                    }

                    @Override
                    public void onAdFailedToLoad(@NonNull LoadAdError adError) {
                        callback.onResult(false);
                        loadedRewarded.remove(id);
                    }
                }
        );
    }

    public void releaseRewarded(String id) {
        loadedRewarded.remove(id);
    }

    public void showRewarded(String id, final RewardedShowCallback callback, final RewardedRewardCallback rewardCallback) {
        RewardedAd rewardedAd = loadedRewarded.get(id);
        if (rewardedAd == null) {
            callback.onResult(0); // Ad not loaded
            return;
        }

        rewardedAd.setFullScreenContentCallback(new FullScreenContentCallback() {
            @Override
            public void onAdFailedToShowFullScreenContent(@NonNull AdError adError) {
                callback.onResult(0); // Failed to show
                loadedRewarded.remove(id);
            }

            @Override
            public void onAdDismissedFullScreenContent() {
                callback.onResult(1); // Dismissed
                loadedRewarded.remove(id);
            }

            @Override
            public void onAdShowedFullScreenContent() {
                callback.onResult(2); // Showed
            }

            @Override
            public void onAdImpression() {
                callback.onResult(3); // Impression
            }

            @Override
            public void onAdClicked() {
                callback.onResult(4); // Clicked
            }
        });

        rewardedAd.show(
                activity, rewardItem -> rewardCallback.onReward(rewardItem.getAmount(), rewardItem.getType()));
    }
}

File: plugin/src/main/java/org/godot/admob/lite/AdLogic.kt
package org.godot.admob.lite

import android.app.Activity
import android.view.Gravity
import android.view.ViewGroup
import android.widget.FrameLayout
import com.google.android.gms.ads.AdError
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.AdSize
import com.google.android.gms.ads.AdView
import com.google.android.gms.ads.FullScreenContentCallback
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.MobileAds
import com.google.android.gms.ads.interstitial.InterstitialAd
import com.google.android.gms.ads.interstitial.InterstitialAdLoadCallback
import com.google.android.gms.ads.rewarded.RewardedAd
import com.google.android.gms.ads.rewarded.RewardedAdLoadCallback

class AdLogic(val act: Activity, val container: FrameLayout) {
	// REQUIRED TO ADD THE APPLICATION ID TO THE MANIFEST https://developers.google.com/admob/android/privacy#add_the_application_id
	val loadedBanners = mutableMapOf<String, AdView>()
	val loadedInterstitials = mutableMapOf<String, InterstitialAd>()
	val loadedRewarded = mutableMapOf<String, RewardedAd>()

	fun initialize() {
		MobileAds.initialize(act)
	}

	//	Banner Logic

	fun loadBanner(id: String, size: Int, position: Int) {
		destroyBanner(id)

		val newBanner = AdView(act)
		if (BuildConfig.DEBUG) {
			newBanner.adUnitId = "ca-app-pub-3940256099942544/6300978111"
		} else {
			newBanner.adUnitId = id
		}

		// Set the AdSize based on the 'size' parameter
		if (size > 40) {
			newBanner.setAdSize(AdSize.getCurrentOrientationAnchoredAdaptiveBannerAdSize(act, size))
		} else {
			when (size) {
				0 -> newBanner.setAdSize(AdSize.BANNER)
				1 -> newBanner.setAdSize(AdSize.LARGE_BANNER)
				2 -> newBanner.setAdSize(AdSize.MEDIUM_RECTANGLE)
				3 -> newBanner.setAdSize(AdSize.FULL_BANNER)
				4 -> newBanner.setAdSize(AdSize.LEADERBOARD)
			}
		}

		// Determine the gravity based on the 'position' parameter
		val gravity = when (position) {
			0 -> Gravity.TOP or Gravity.CENTER_HORIZONTAL
			1 -> Gravity.CENTER
			2 -> Gravity.BOTTOM or Gravity.CENTER_HORIZONTAL
			else -> Gravity.TOP or Gravity.CENTER_HORIZONTAL // Default to top if an invalid position is provided
		}

		// Create a new FrameLayout.LayoutParams with the determined gravity
		val layoutParams = FrameLayout.LayoutParams(
			FrameLayout.LayoutParams.MATCH_PARENT, // Width
			FrameLayout.LayoutParams.WRAP_CONTENT, // Height
			gravity
		)

		// Add the new banner to the container with the specified layout parameters
		container.addView(newBanner, layoutParams)

		val adRequest = AdRequest.Builder().build()
		newBanner.loadAd(adRequest)
		loadedBanners[id] = newBanner
	}

	fun destroyBanner(id: String) {
		// Remove banner from view hierarchy.
		val parentView = loadedBanners[id]?.parent
		if (parentView is ViewGroup) {
			parentView.removeView(loadedBanners[id])
		}
		loadedBanners[id]?.destroy()
		loadedBanners.remove(id)
	}

	//	Interstitial Logic

	fun loadInterstitial(id: String, callback: (Boolean) -> Unit) {
		var adId = "ca-app-pub-3940256099942544/1033173712"
		if (!BuildConfig.DEBUG) {
			adId = id
		}
		InterstitialAd.load(
			act,
			adId,
			AdRequest.Builder().build(),
			object : InterstitialAdLoadCallback() {
				override fun onAdLoaded(ad: InterstitialAd) {
					callback(true)
					loadedInterstitials[id] = ad
				}

				override fun onAdFailedToLoad(adError: LoadAdError) {
					callback(false)
					loadedInterstitials.remove(id)
				}
			},
		)
	}

	fun releaseInterstitial(id: String) {
		loadedInterstitials.remove(id)
	}

	fun showInterstitial(id: String, callback: (Int) -> Unit) {
		if (loadedInterstitials[id] == null) {
			callback(0)
		}
		loadedInterstitials[id]?.fullScreenContentCallback = object : FullScreenContentCallback() {
			override fun onAdFailedToShowFullScreenContent(adError: AdError) {
				callback(0)
				loadedInterstitials.remove(id)
			}

			override fun onAdDismissedFullScreenContent() {
				callback(1)
				loadedInterstitials.remove(id)
			}

			override fun onAdShowedFullScreenContent() {
				callback(2)
			}

			override fun onAdImpression() {
				callback(3)
			}

			override fun onAdClicked() {
				callback(4)
			}
		}
		loadedInterstitials[id]?.show(act)
	}

	//	Rewarded Logic

	fun loadRewarded(id: String, callback: (Boolean) -> Unit) {
		var adId = "ca-app-pub-3940256099942544/5224354917"
		if (!BuildConfig.DEBUG) {
			adId = id
		}

		RewardedAd.load(
			act,
			adId,
			AdRequest.Builder().build(),
			object : RewardedAdLoadCallback() {
				override fun onAdLoaded(ad: RewardedAd) {
					callback(true)
					loadedRewarded[id] = ad
				}

				override fun onAdFailedToLoad(adError: LoadAdError) {
					callback(false)
					loadedRewarded.remove(id)
				}
			},
		)
	}

	fun releaseRewarded(id: String) {
		loadedRewarded.remove(id)
	}

	fun showRewarded(id: String, callback: (Int) -> Unit, rewardCallback: (Int, String) -> Unit) {
		if (loadedRewarded[id] == null) {
			callback(0)
		}
		loadedRewarded[id]?.fullScreenContentCallback = object : FullScreenContentCallback() {
			override fun onAdFailedToShowFullScreenContent(adError: AdError) {
				callback(0)
				loadedRewarded.remove(id)
			}

			override fun onAdDismissedFullScreenContent() {
				callback(1)
				loadedRewarded.remove(id)
			}

			override fun onAdShowedFullScreenContent() {
				callback(2)
			}

			override fun onAdImpression() {
				callback(3)
			}

			override fun onAdClicked() {
				callback(4)
			}
		}

		loadedRewarded[id]?.show(
			act,
		) { rewardItem ->
			rewardCallback(rewardItem.amount, rewardItem.type)
		}
	}
}